File: codeforces/1600~1699/1679/d.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  int n = 0, m = 0;
  i64 k = 0;
  cin >> n >> m >> k;

  vector<int> a(n+1);
  for (int i = 1; i <= n; ++i) cin >> a[i];

  vector<vector<int>> adj(n+1);
  for (int i = 0; i < m; ++i) {
    int u = 0, v = 0;
    cin >> u >> v;
    adj[u].push_back(v);
  }

  vector<bool> visited(n+1);
  vector<int> nodes;
  function<void(int)> top_sort = [&](int node) {
    if (visited[node]) {
      return;
    }
    visited[node] = true;
    for (int next_node : adj[node]) {
      top_sort(next_node);
    }
    nodes.push_back(node);
  };

  for (int node = 1; node <= n; ++node) {
    if (visited[node]) {
      continue;
    }
    top_sort(node);
  }

  reverse(nodes.begin(), nodes.end());

  vector<i64> memo(n+1, -1);
  function<i64(int,int)> dfs = [&](int node, int max_val) -> i64 {
    if (a[node] > max_val) {
      return 0;
    }
    if (memo[node] >= 0) {
      return memo[node];
    }
    if (visited[node]) {
      // cout << "found cycle at " << node << endl;
      // found a cycle
      return k;
    }
    visited[node] = true;
    i64 ans = 1;
    for (int next_node : adj[node]) {
      ans = max(ans, dfs(next_node, max_val) + 1);
    }
    visited[node] = false;
    return memo[node] = ans;
  };

  auto path_found = [&](int max_val) -> bool {
    fill(visited.begin(), visited.end(), false);
    fill(memo.begin(), memo.end(), -1);
    for (int node : nodes) {
      // if (visited[node]) {
      //   continue;
      // }

      i64 len = dfs(node, max_val);
      if (len >= k) {
        return true;
      }
    }
    return false;
  };

  int l = *min_element(a.begin(), a.end());
  int r = *max_element(a.begin(), a.end());
  int ans = -1;
  while (l <= r) {
    int mid = l + (r-l) / 2;
    if (path_found(mid)) {
      // cout << "path found for " << mid << endl;
      ans = mid;
      r = mid-1;
    } else {
      l = mid+1;
    }
  }

  cout << ans << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  solve();
  return 0;
}

File: codeforces/1600~1699/1679/d.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;
use std::cmp::max;
 
fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}
 
fn top_sort(node: usize, adj: &Vec<Vec<i32>>, visited: &mut Vec<bool>, nodes: &mut Vec<i32>) {
    if visited[node] {
        return;
    }
    visited[node] = true;
 
    for next_node in &adj[node] {
        top_sort(*next_node as usize, adj, visited, nodes);
    }
    nodes.push(node as i32);
}
 
fn dfs(node: usize, max_val: i32, k: i64, a: &Vec<i32>, adj: &Vec<Vec<i32>>, visited: &mut Vec<bool>, memo: &mut Vec<i64>) -> i64 {
    if a[node] > max_val {
        return 0;
    }
 
    if memo[node] >= 0 {
        return memo[node];
    }
 
    if visited[node] {
        return k;
    }
    visited[node] = true;
    let mut ans: i64 = 1;
    for next_node in &adj[node] {
        ans = max(ans, dfs(*next_node as usize, max_val, k, a, adj, visited, memo) + 1);
    }
    visited[node] = false;
    memo[node] = ans;
    ans
}
 
fn path_found(max_val: i32, k: i64, nodes: &Vec<i32>, a: &Vec<i32>, adj: &Vec<Vec<i32>>) -> bool {
    let n = a.len();
    let mut visited: Vec<bool> = vec![false; n+1];
    let mut memo: Vec<i64> = vec![-1; n+1];
 
    for node in nodes {
        let len = dfs(*node as usize, max_val, k, a, adj, &mut visited, &mut memo);
        if len >= k {
            return true;
        }
    }
    false
}
 
fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());
 
    let inputs = read_vec::<i64>();
    let n = inputs[0] as usize;
    let m = inputs[1] as usize;
    let k = inputs[2];
 
    let mut a = read_vec::<i32>();
    a.insert(0, 0);
    let mut adj: Vec<Vec<i32>> = vec![Vec::new(); n+1];
    for _i in 0..m {
        let edge = read_vec::<i32>();
        let u = edge[0];
        let v = edge[1];
        adj[u as usize].push(v);
    }
 
    let mut nodes: Vec<i32> = Vec::new();
    let mut visited: Vec<bool> = vec![false; n+1];
 
    for node in 1..=n {
        if !visited[node] {
            top_sort(node, &adj, &mut visited, &mut nodes);
        }
    }
 
    nodes.reverse();
 
    let mut l = *a.iter().min().unwrap();
    let mut r = *a.iter().max().unwrap();
    let mut ans = -1;
    while l <= r {
        let mid = l + (r-l) / 2;
        if path_found(mid, k, &nodes, &a, &adj) {
            ans = mid;
            r = mid-1;
        } else {
            l = mid+1;
        }
    }
 
    writeln!(writer, "{}", ans).ok();
}
 
fn main() {
    solve();
}
